peribolos: Allow renaming repositories

petr-muller opened the pull request:
Builds on https://github.com/kubernetes/test-infra/pull/14885 which hopefully close to merge. Contains only one additional commit over https://github.com/kubernetes/test-infra/pull/14885.

Allows renaming repositories, using the `previously:` pattern used in other parts in peribolos and elsewhere.

/cc @fejta @stevekuznetsov

fejta commented on prow/cmd/peribolos/main.go:
```diff
@@ -877,13 +882,18 @@ func validateRepos(repos map[string]org.Repo) error {
-			continue
+	for wantName, repo := range repos {
+		toCheck := []string{wantName}
+		toCheck = append(toCheck, repo.Previously...)
```
nit: I believe `toCheck := append([]string{wantName}, repo.Previously...)` will work here

fejta commented on prow/cmd/peribolos/main.go:
```diff
@@ -893,7 +903,73 @@ func validateRepos(repos map[string]org.Repo) error {
+	return errs
+}
+
+func repoExists(repos map[string]github.Repo, name string, previousNames []string) (*github.Repo, bool) {
```
Do we want to support changing the casing? Do we already?

petr-muller replied:
Good point. We don't, but not intentionally: it just gets detected as a duplicate name and error is issued,

fejta commented on prow/cmd/peribolos/main.go:
```diff
@@ -893,7 +903,73 @@ func validateRepos(repos map[string]org.Repo) error {
+	names := []string{name}
+	names = append(names, previousNames...)
+	for _, name := range names {
+		if repo, exists := repos[strings.ToLower(name)]; exists {
```
What do we want to do when both `C` and its previous name `P` exist? Should we archive `P`?

petr-muller commented:
@fejta Rebased and addressed your comments (added tests for these cases, too).

- Renames that are just a change of casing (`repo`->`Repo`) are now supported
- Cases when two repos exist for both current and some previous name, or when two different repos exist for two previous names of a repo are detected and reported as errors.

fejta commented on prow/cmd/peribolos/main_test.go:
```diff
@@ -2773,6 +2773,84 @@ func TestConfigureRepos(t *testing.T) {
+			description: "renaming a repo by just changing case is successful",
+			orgConfig: org.Config{
+				Repos: map[string]org.Repo{
+					"repo": {Previously: []string{"REPO"}},
```
I actually think that if I declare `Repo` and the current name is `rEpO` then we should rename it to `Repo` -- without requiring a previously.

petr-muller replied:
Makes sense, will fix!

petr-muller replied:
It actually worked like this in the PR. I just added a test for it.
